File: assets/proof/sample.py
"""Proof corpus — Python.

Exercises: decorators, f-strings, type hints, dunder methods, comprehensions,
keyword arguments, `self`, builtins.
"""
from __future__ import annotations

import re
from dataclasses import dataclass, field
from enum import Enum
from functools import cached_property

CALLSIGN = re.compile(r"^[A-Z]{2}-\d{3}$")
CEILING_FT: int = 65_000


class Mode(str, Enum):
    NORMAL = "normal"
    INSERT = "insert"
    VISUAL = "visual"


@dataclass(slots=True)
class Airframe:
    callsign: str
    thrust: float = 0.82
    telemetry: dict[str, float] = field(default_factory=dict)

    def __post_init__(self) -> None:
        if not CALLSIGN.match(self.callsign):
            raise ValueError(f"bad callsign: {self.callsign!r}")

    @cached_property
    def modes(self) -> tuple[Mode, ...]:
        return tuple(m for m in Mode if m is not Mode.VISUAL)

    def climb(self, feet: int, *, strict: bool = True) -> int:
        if strict and feet > CEILING_FT:
            raise OverflowError(f"{feet} exceeds ceiling {CEILING_FT}")
        self.telemetry["alt"] = float(feet)
        return feet

File: assets/proof/sample.rs
//! Proof corpus — Rust.
//! Exercises: lifetimes, attributes, macros, traits, generics, match, `self`.

use std::collections::HashMap;
use std::fmt::{self, Display};

const CEILING_FT: u32 = 65_000;

#[derive(Debug, Clone, PartialEq)]
pub enum Mode { Normal, Insert, Visual { lines: bool } }

pub trait Annunciator {
    fn label(&self) -> &'static str;
    fn lit(&self) -> bool { true }
}

impl Annunciator for Mode {
    fn label(&self) -> &'static str {
        match self {
            Mode::Normal => "NORMAL",
            Mode::Insert => "INSERT",
            Mode::Visual { lines: true } => "V-LINE",
            Mode::Visual { .. } => "VISUAL",
        }
    }
}

#[derive(Default)]
pub struct Airframe<'a> {
    pub callsign: &'a str,
    pub thrust: f64,
    telemetry: HashMap<String, f64>,
}

impl<'a> Airframe<'a> {
    pub fn new(callsign: &'a str) -> Self {
        Self { callsign, thrust: 0.82, telemetry: HashMap::new() }
    }

    pub fn climb(&mut self, feet: u32) -> Result<u32, String> {
        if feet > CEILING_FT {
            return Err(format!("{} exceeds ceiling {CEILING_FT}", feet));
        }
        self.telemetry.insert("alt".into(), f64::from(feet));
        Ok(feet)
    }
}

impl Display for Mode {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result { write!(f, "{}", self.label()) }
}
